client/consensus/grandpa/authorities.go

```diff
@@ -0,0 +1,908 @@
+		setID:       setID,
+		blockNumber: blockNumber,
+	})
+}
+
+type authoritySetChangeID scale.VaryingDataType
+
+// Set will set a VaryingDataTypeValue using the underlying VaryingDataType
+func (asc *authoritySetChangeID) Set(val scale.VaryingDataTypeValue) (err error) {
+	vdt := scale.VaryingDataType(*asc)
+	err = vdt.Set(val)
+	if err != nil {
+		return
+	}
+	*asc = authoritySetChangeID(vdt)
+	return
+}
+
+// Value will return value from underying VaryingDataType
+func (asc *authoritySetChangeID) Value() (val scale.VaryingDataTypeValue, err error) {
+	vdt := scale.VaryingDataType(*asc)
+	return vdt.Value()
+}
+
+func newAuthoritySetChangeID[N constraints.Unsigned]() authoritySetChangeID {
+	vdt, err := scale.NewVaryingDataType(latest{}, set[N]{}, unknown{})
+	if err != nil {
+		panic(err)
+	}
+	return authoritySetChangeID(vdt)
+}
+
+type latest struct{}
+
+func (latest) Index() uint {
+	return 0
+}
+
+type set[N constraints.Unsigned] struct {
+	inner setIDNumber[N]
+}
+
+func (set[N]) Index() uint {
+	return 1
+}
+
+type unknown struct{}
+
+func (unknown) Index() uint {
+	return 2
+}
+
+// Three states that can be returned: Latest, Set (tuple), Unknown
+func (asc *AuthoritySetChanges[N]) getSetID(blockNumber N) (authSetChangeID authoritySetChangeID, err error) {
+	if asc == nil {
+		return authSetChangeID, fmt.Errorf("getSetID: authSetChanges is nil")
+	}
+	authSetChangeID = newAuthoritySetChangeID[N]()
+	authSet := *asc
+	last := authSet[len(authSet)-1]
+	if last.blockNumber < blockNumber {
+		err = authSetChangeID.Set(latest{})
+		if err != nil {
+			return authSetChangeID, err
+		}
+		return authSetChangeID, nil
+	}
+
+	idx, _ := slices.BinarySearchFunc(
+		authSet,
+		blockNumber,
+		func(a setIDNumber[N], b N) int {
+			switch {
+			case a.blockNumber == b:
+				return 0
+			case a.blockNumber > b:
+				return 1
+			case a.blockNumber < b:
+				return -1
+			default:
+				panic("huh?")
+			}
+		},
+	)
+	if idx < len(authSet) {
+		authChange := authSet[idx]
+
+		// if this is the first index but not the first set id then we are missing data.
+		if idx == 0 && authChange.setID != 0 {
+			err = authSetChangeID.Set(unknown{})
+			if err != nil {
+				return authSetChangeID, err
+			}
+			return authSetChangeID, nil
+		}
+		err = authSetChangeID.Set(set[N]{
+			authChange,
+		})
+		if err != nil {
+			return authSetChangeID, err
+		}
+		return authSetChangeID, nil
+	}
+
+	err = authSetChangeID.Set(unknown{})
+	if err != nil {
+		return authSetChangeID, err
+	}
+	return authSetChangeID, nil
+}
+
+func (asc *AuthoritySetChanges[N]) insert(blockNumber N) {
+	var idx int
+	if asc == nil {
+		panic("authority set changes must be initialised")
+	} else {
+		idx, _ = slices.BinarySearchFunc(
+			*asc,
+			blockNumber,
+			func(a setIDNumber[N], b N) int {
+				switch {
+				case a.blockNumber == b:
+					return 0
+				case a.blockNumber > b:
+					return 1
+				case a.blockNumber < b:
+					return -1
+				default:
+					panic("huh?")
```

Comment:
looks like all possible comparisons are covered so I think you can remove the `default` case

Reply:
same here